laundry/views.py

```diff
@@ -0,0 +1,167 @@
+        try:
+            (room, snapshots) = self.get_snapshot_info(hall_id)
+        except ValueError:
+            return Response({"error": "Invalid hall id passed to server."}, status=404)
+
+        # [0]: available washers, [1]: available dryers, [2]: total number of LaundrySnapshots
+        data = [(0, 0, 0)] * 27
+
+        # used calculate the start and end dates
+        min_date = timezone.localtime()
+        max_date = timezone.localtime() - datetime.timedelta(days=30)
+
+        for snapshot in snapshots:
+            date = snapshot.date.astimezone()
+
+            if date < min_date:
+                min_date = date
+
+            if date > max_date:
+                max_date = date
+
+            hour = date.hour
+
+            # accounts for the 3 hours on the next day
+            if (
+                calendar.day_name[timezone.localtime().weekday()]
+                == calendar.day_name[(date - datetime.timedelta(days=1)).weekday()]
+            ):
+                hour = date.hour + 24
+
+            # adds total number of available washers and dryers
+            data[hour] = (
+                data[hour][0] + snapshot.available_washers,
+                data[hour][1] + snapshot.available_dryers,
+                data[hour][2] + 1,
+            )
+
+        # collects total washers and dryers
+        try:
+            total_washers = snapshots.first().total_washers
+            total_dryers = snapshots.first().total_dryers
+        except AttributeError:
+            total_washers = 0
+            total_dryers = 0
+
+        return Response(
+            {
+                "hall_name": room.name,
+                "location": room.location,
+                "day_of_week": calendar.day_name[timezone.localtime().weekday()],
+                "start_date": min_date.date(),
+                "end_date": max_date.date(),
+                "washer_data": {x: self.safe_division(data[x][0], data[x][2]) for x in range(27)},
+                "dryer_data": {x: self.safe_division(data[x][1], data[x][2]) for x in range(27)},
+                "total_number_of_washers": total_washers,
+                "total_number_of_dryers": total_dryers,
+            }
+        )
+
+
+class Preferences(APIView):
+    """
+    GET: returns list of a User's laundry preferences
+
+    POST: updates User laundry preferences by clearing past preferences
+    and resetting them with request data
+    """
+
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request):
+
+        preferences = request.user.profile.laundry_preferences.all()
+
+        # returns all hall_ids in a person's preferences
+        return Response({"rooms": preferences.values_list("hall_id", flat=True)})
+
+    def post(self, request):
+
+        profile = request.user.profile
+
+        # clears all previous preferences in many-to-many
+        profile.laundry_preferences.clear()
+
+        hall_ids = request.data["rooms"]
+
+        for hall_id in hall_ids:
```

Comment:
This loop should probably be in a try/catch that returns a 400 if something goes wrong. Normally DRF would handle that for us, but since we're not using serializers to do this update we have to do it manually. Down the line, we should probably revisit this to make it more DRF-y